IEEE-PSH/hackpsh-web
# Create Base User Record on Redirect to Onboarding

# Description

In the instance of a brand new user, they will get redirected to /onboarding/** and will be locked in that path until their db_record of user_onboarding_complete returns true. As they get redirected to onboarding with a valid session, they will hit the middleware, and where it will check if the current user has a record within the database, and if not, procede to invoke a serverTRPC procedure which will trigger a database function to create a base user record. This will be the record that will be used from here on-out for the brand new user throughout onboarding and the rest of the platform.

# Notes
As this is in the middleware, this'd mean all protected routes would invoke a db_record lookup (does_user_exist & is_onboarding_complete) for every single route traversal. We should try to find a way to optimize this to only happen on sign-in, (/api/auth/callback) however, invoking serverTRPC there under the presumption of a protected record does not appear to work. The solution would be to use server-side helpers, however, we run into #33 or to turn create_user into a public procedure but that opens up the possibility of ddos attacks onto our DB (Allowing mutations on our DB on a public route is bad, because people can just hit the endpoint themselves).

I've definitely also skimped a little bit on the amount of error reporting and try catch blocks throughout this implementation. As of right now, we don't have an easy way to pipe out errors into a system. This majorly due to how large the middleware function is starting to get, and I need opinions here on creating a pattern that simplifies this function, whether that's extracting this logic else-where or etc. 

# Testing Guidelines

* Login via localhost and see that protected routes force-redirect you to /onboarding
* View in app_user_profile table for the Base User Record being created
